src/DotLiquid/Util/StrFTime.cs

```diff
@@ -18,85 +20,197 @@ public static class StrFTime
+        /// - `%Z` (Not standard) - DotLiquid returns an offset (+00:00), Ruby states this should return a timezone description (e.g. 'GMT')
+        /// <see href="https://help.shopify.com/themes/liquid/filters/additional-filters#date"/>
+        /// <see href="https://ruby-doc.org/core-3.0.0/Time.html#method-i-strftime"/>
+        /// </summary>
+        /// <param name="dateTime"/>
+        /// <param name="pattern"/>
+        /// <returns>a string version of dateTime matching pattern.</returns>
         public static string ToStrFTime(this DateTime dateTime, string pattern)
         {
-            string output = "";
-
-            int n = 0;
-
-            while (n < pattern.Length)
-            {
-                string s = pattern.Substring(n, 1);
-
-                if (n + 1 >= pattern.Length)
-                    output += s;
-                else
-                    output += s == "%"
-                        ? Formats.ContainsKey(pattern.Substring(++n, 1)) ? Formats[pattern.Substring(n, 1)].Invoke(dateTime) : "%" + pattern.Substring(n, 1)
-                        : s;
-                n++;
-            }
-
-            return output;
+            return Regex.Replace(pattern, @"%(?<flag>[-_0^:#])*(?<width>[1-9][0-9]*)?(?<directive>[a-zA-Z%\+])",
+                x => StrFTimeMatchEvaluator(
+                    x.Groups[0].Value,
+                    x.Groups["flag"].Captures.Cast<Capture>().Select(y => y.Value).ToList(),
+                    x.Groups["width"].Captures.Cast<Capture>().Select(y => (int?)Convert.ToInt32(y.Value)).FirstOrDefault(),
+                    x.Groups["directive"].Captures.Cast<Capture>().Select(y => y.Value).FirstOrDefault(),
+                    dateTime
+                    ));
         }
 
         /// <summary>
         /// Formats a date using a ruby date format string
         /// </summary>
-        /// <param name="dateTime"></param>
+        /// <param name="dateTimeOffset"></param>
         /// <param name="pattern"></param>
         /// <returns></returns>
-        public static string ToStrFTime(this DateTimeOffset dateTime, string pattern)
+        public static string ToStrFTime(this DateTimeOffset dateTimeOffset, string pattern)
+        {
+            return Regex.Replace(pattern, @"%(?<flag>[-_0^:#])*(?<width>[1-9][0-9]*)?(?<directive>[a-zA-Z%\+])",
+                x => StrFTimeMatchEvaluator(
+                    x.Groups[0].Value,
+                    x.Groups["flag"].Captures.Cast<Capture>().Select(y => y.Value).ToList(),
+                    x.Groups["width"].Captures.Cast<Capture>().Select(y => (int?)Convert.ToInt32(y.Value)).FirstOrDefault(),
+                    x.Groups["directive"].Captures.Cast<Capture>().Select(y => y.Value).FirstOrDefault(),
+                    dateTimeOffset
+                    ));
+        }
+
+        private static String StrFTimeMatchEvaluator(String orig, IEnumerable<String> flags, int? width, String directive, object obj)

```

Comment:
The concepts used here such as 'directive' may not be obvious. Best provide XML comment of what the method is doing and what the parameters are.

Reply:
Added XML comment.